client/client-lib/src/mint/backup.rs

```diff
@@ -0,0 +1,218 @@
+
+        self.upload_ecash_backup(backup).await?;
+
+        Ok(())
+    }
+
+    pub async fn download_ecash_backup_from_federation(&self) -> Result<PlaintextEcashBackup> {
+        let encrypted = self
+            .context
+            .api
+            .download_ecash_backup(&self.get_derived_backup_signing_key().x_only_public_key().0)
+            .await?;
+
+        EcashBackup(encrypted).decrypt_with(&self.get_derived_backup_encryption_key())
+    }
+
+    /// Static version of [`Self::get_derived_backup_encryption_key`] for testing without creating whole `MintClient`
+    fn get_derived_backup_encryption_key_static(secret: &DerivableSecret) -> aead::LessSafeKey {
+        // TODO: Do we need that one derivation level? This key is already derived for the mint itself, and internally another kdf will be done with key type tag.
+        aead::LessSafeKey::new(
+            secret
+                .child_key(ChildId(BACKUP_CHILD_ID))
+                .to_chacha20_poly1305_key(),
+        )
+    }
+
+    /// Static version of [`Self::get_derived_backup_secret_key`] for testing without creating whole `MintClient`
+    fn get_derived_backup_signing_key_static(secret: &DerivableSecret) -> secp256k1_zkp::KeyPair {
+        // TODO: Do we need that one derivation level? This key is already derived for the mint itself, and internally another kdf will be done with key type tag.
+        secret
+            .child_key(ChildId(BACKUP_CHILD_ID))
+            .to_secp_key(&Secp256k1::<secp256k1::SignOnly>::gen_new())
+    }
+
+    fn get_derived_backup_encryption_key(&self) -> aead::LessSafeKey {
+        Self::get_derived_backup_encryption_key_static(&self.secret)
+    }
+
+    fn get_derived_backup_signing_key(&self) -> secp256k1::KeyPair {
+        Self::get_derived_backup_signing_key_static(&self.secret)
+    }
+
+    async fn prepare_plaintext_ecash_backup(&self) -> PlaintextEcashBackup {
+        let mut dbtx = self.start_dbtx();
+
+        let notes = self.get_available_notes(&mut dbtx);
+        let last_idx = self.get_last_note_index(&mut dbtx);
+
+        PlaintextEcashBackup { notes, last_idx }
+    }
+
+    async fn prepare_ecash_backup(&self) -> Result<EcashBackup> {
+        let plaintext = self.prepare_plaintext_ecash_backup().await;
+        plaintext.encrypt_to(&self.get_derived_backup_encryption_key())
+    }
+
+    async fn upload_ecash_backup(&self, backup: EcashBackup) -> Result<()> {
+        let backup_request = backup.into_backup_request(&self.get_derived_backup_signing_key())?;
+        self.context
+            .api
+            .upload_ecash_backup(&backup_request)
+            .await?;
+        Ok(())
+    }
+}
+
+/// Snapshot of a ecash state (notes)
+///
+/// Used to speed up and improve privacy of ecash recovery,
+/// by avoiding scanning the whole history.
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct PlaintextEcashBackup {
+    notes: TieredMulti<SpendableNote>,
+    last_idx: NoteIndex,
+}
```

Comment:
Should probably also include the current epoch?

Reply:
I thought so initially, but ... I don't see why would it matter anymore.

Reply:
There isn't a really good `wall clock time <> epoch` mapping, so the timestamp isn't a replacement for the epoch imo.

We do the regular backup so we only have to download the epoch history from the last point of having mad a backup. So we definitely want to know where to begin downloading.

Reply:
> We do the regular backup so we only have to download the epoch history from the last point of having mad a backup. 

Isn't it going to be note-index based anyway? As in we generate every nonces and ask the federation about it?

Or are we going to go through all transaction history from a given epoch instead? Note-based approach seems lighter and simpler, but compromises on privacy.

Reply:
> Or are we going to go through all transaction history from a given epoch instead? Note-based approach seems lighter and simpler, but compromises on privacy.

Asking for specific e-cash notes would be pretty bad privacy-wise. Downloading the entire log, but only for a short period of time (since last daily backup) seems like a good compromise until we have some fancy PIR scheme.